sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2007,7 +2007,14 @@ case class Concat(children: Seq[Expression]) extends Expression {
     }
   }
 
-  override def dataType: DataType = children.map(_.dataType).headOption.getOrElse(StringType)
+  override def dataType: DataType = {
+    val dataTypes = children.map(_.dataType)
+    dataTypes.headOption.map {
+      case ArrayType(et, _) =>
+        ArrayType(et, dataTypes.exists(_.asInstanceOf[ArrayType].containsNull))
```

Comment:
do we support array of array in concat?

Reply:
Yes, it should work (see a [test](https://github.com/ueshin/apache-spark/blob/30d5aed41085cfb82e3da43099adf972b953ece8/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala#L951) for it). Did we miss anything?

Reply:
then shall we fix the `containNull` for the inner array?

Reply:
yeah, + ```valueContainsNull``` for ```MapType``` and ```nullable``` for ```StructField```

Reply:
I guess the inner nullabilities are coerced during type-coercion? If the inner nullabilities are different, type coercion adds casts and they will remain.

Reply:
@ueshin For ```Concat```, ```Coalesce```, etc.  it seems to be that case since a coercion rule is executed if there is any nullability difference on any level of nesting. But it's not the case of ```CaseWhenCoercion``` rule, since ```sameType``` method is used for comparison.

I'm wondering if the goal is to avoid generation of extra ```Cast``` expressions, shouldn't other coercion rules utilize ```sameType``` method as well? Let's assume that the result of ```concat``` is subsequently used by ```flatten```, wouldn't  it lead to generation of extra null safe checks as mentioned [here](https://github.com/apache/spark/pull/21704#discussion_r200110924)?

Reply:
Please ignore the part of my previous comment regarding ```flatten``` function. The output data type of ```concat```, etc. will be the same regardless what resolves ```null``` flags.